stonesoup/hypothesiser/mfa.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+from .base import Hypothesiser
+from ..base import Property
+from ..types.multihypothesis import MultipleHypothesis
+from ..types.prediction import TaggedWeightedGaussianStatePrediction
+
+
+class MFAHypothesiser(Hypothesiser):
+    """Multi-Frame Assignment Hypothesiser based on an underlying Hypothesiser
+    Generates a list of SingleHypotheses pertaining to individual component-detection hypotheses
+    """
+
+    hypothesiser: Hypothesiser = Property(
+        doc="Underlying hypothesiser used to generate detection-target pairs")
+
+    def hypothesise(self, track, detections, timestamp):
+        """Form hypotheses for associations between Detections and a given track.
+
+        Parameters
+        ----------
+        track: :class:`~.Track`
+            The track object to hypothesise on
+        detections : list of :class:`Detection`
+            Retrieved measurements
+        timestamp : datetime
+            Time of the detections/predicted states
+        Returns
+        -------
+        : :class:`~.MultipleHypothesis`
+            A container of :class:`~SingleProbabilityHypothesis` objects, pertaining to individual
+            component-detection hypotheses
+        """
+
+        # Check to make sure all detections are obtained from the same time
+        timestamps = set([detection.timestamp for detection in detections])
+        if len(timestamps) > 1:
+            raise ValueError("All detections must have the same timestamp")
+
+        hypotheses = list()
+        detections_list = list(detections)
+        for component in track.state.components:
```

Comment:
This line is breaking things when I try to run an example.  I can't find any mention of `state.components` anywhere, and when I try to run something like `print(state.components)` for a valid State or Gaussian instance I get an `AttributeError`.  Also, there are compatibility issues as to which underlying hypothesiser is used: if we initialise using `DistanceHypothesiser`, for example, this class' `hypothesise` method takes the whole `Track` as a parameter.  I'm not sure which underlying hypothesiser is most appropriate, as I'm still getting to grips with how the algorithm functions, but some kind of way of dealing with someone attempting to use the wrong one is probably needed, or some guidance in the documentation.

Reply:
I think you maybe meant `for component in track.states:`

Reply:
This seems to assume that `track.state` is a `GaussianMixture`, which does have a `components` attribute, and I think later code assumes it consists of `TaggedWeightedGaussianState` objects. I could perhaps put a couple of `assert isinstance(...)` statements here to assure that, but that would go against Python's duck typing philosophy (if the user came up with their own classes with the same interface they would be prevented from passing that in). Maybe a better solution would be to wrap relevant statements in `try: ... except AttributeError:` and re-throw the exception with a more useful error message – would that suit?

Reply:
You could use `hasattr(obj, "components")`, but I think just adding a note to documentation is all that's needed